common/numerated/src/iterators.rs

```diff
@@ -0,0 +1,330 @@
+    pub(crate) iter1: I,
+    pub(crate) iter2: I,
+    pub(crate) interval1: Option<Interval<T>>,
+    pub(crate) interval2: Option<Interval<T>>,
+}
+
+impl<T: Numerated, I: Iterator<Item = Interval<T>>> Iterator for DifferenceIterator<T, I> {
+    type Item = Interval<T>;
+
+    fn next(&mut self) -> Option<Self::Item> {
+        loop {
+            let interval1 = if let Some(interval1) = self.interval1 {
+                interval1
+            } else {
+                let interval1 = self.iter1.next()?;
+                self.interval1 = Some(interval1);
+                interval1
+            };
+
+            let interval2 = if let Some(interval2) = self.interval2 {
+                interval2
+            } else if let Some(interval2) = self.iter2.next() {
+                interval2
+            } else {
+                self.interval1 = None;
+                return Some(interval1);
+            };
+
+            if interval2.end() < interval1.start() {
+                self.interval2 = None;
+                continue;
+            }
+
+            self.interval2 = Some(interval2);
+
+            if interval1.end() < interval2.start() {
+                self.interval1 = None;
+                return Some(interval1);
+            } else {
+                if let Some(new_start) = interval2.end().inc_if_lt(interval1.end()) {
+                    self.interval1 = Interval::new(new_start, interval1.end());
+                    debug_assert!(self.interval1.is_some(), "`T: Numerated` impl error");
+                } else if interval1.end() == interval2.end() {
+                    self.interval1 = None;
+                    self.interval2 = None;
+                } else {
+                    self.interval1 = None;
+                }
+
+                if let Some(new_end) = interval2.start().dec_if_gt(interval1.start()) {
+                    let res = Interval::new(interval1.start(), new_end);
+                    debug_assert!(res.is_some(), "`T: Numerated` impl error");
+                    return res;
+                } else {
+                    continue;
+                }
+            }
+        }
+    }
+}
+
+/// Helper struct to iterate over voids in tree.
+///
+/// See also [`IntervalsTree::voids`].
+pub struct VoidsIterator<T: Numerated, I: Iterator<Item = Interval<T>>> {
+    pub(crate) inner: Option<(I, Interval<T>)>,
+}
+
+impl<T: Numerated, I: Iterator<Item = Interval<T>>> Iterator for VoidsIterator<T, I> {
+    type Item = Interval<T>;
+
+    fn next(&mut self) -> Option<Self::Item> {
+        let (iter, interval) = self.inner.as_mut()?;
+        if let Some(next) = iter.next() {
+            let (start, end) = next.into_parts();
+
+            // Guaranties by tree: between two intervals always exists void.
+            debug_assert!(interval.start() < start);
+
+            let Some(void_end) = start.dec_if_gt(interval.start()) else {
+                debug_assert!(false, "`T: Numerated` impl error");
```

Comment:
shouldn't we panic in prod for such cases?

Reply:
We, can panic. In rust core you can find both ways when some `T` implementation is incorrect:
1) returns wrong result
2) panics
3) sometimes also return error in result (but it's rare case)

I choose finally in numerated `wrong result` in order to make probability of panics on migration as less as it's possible.

Reply:
done